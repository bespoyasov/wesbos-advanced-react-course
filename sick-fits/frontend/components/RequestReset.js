import React, {Component} from 'react'
import {Mutation} from 'react-apollo'
import gql from 'graphql-tag'

import Form from './styles/Form'
import Error from './ErrorMessage'


const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`


const initialState = {
  email: '',
}


class RequestReset extends Component {
  state = {...initialState}

  updateState = (e) => {
    const {name, value} = e.target
    this.setState(state => ({ [name]: value }))
  }

  render() {
    const {email} = this.state

    return (
      <Mutation 
        mutation={REQUEST_RESET_MUTATION} 
        variables={this.state}>
        {(requestReset, {error, loading, called}) => {
          return (
            <Form method='post' onSubmit={async e => {
              e.preventDefault()
              await requestReset()
            }}>
              <fieldset disabled={loading} aria-busy={loading}>
                <h2>Reset a password</h2>
                <Error error={error} />

                {!error && !loading && called && (
                  <p>Success! Check your email</p>
                )}

                <label>
                  Email
                  <input 
                    type='email' 
                    name='email'
                    placehodler='Email'
                    value={email} 
                    onChange={this.updateState} />
                </label>
      
                <button type='submit'>Submit!</button>
              </fieldset>
            </Form>
          )
        }}
      </Mutation>
    )
  }
}

export default RequestReset