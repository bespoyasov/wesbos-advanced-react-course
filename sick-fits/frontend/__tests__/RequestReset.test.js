import {mount} from 'enzyme'
import wait from 'waait'
import toJSON from 'enzyme-to-json'
import {MockedProvider} from 'react-apollo/test-utils'

import RequestReset, {
  REQUEST_RESET_MUTATION as query,
} from '../components/RequestReset'


const email = 'test@test.com'
const mocks = [{
  request: {
    query, 
    variables: {email}
  },
  result: {
    data: {
      requestReset: {
        __typename: 'message',
        message: 'success',
      }
    }
  }
}]


describe('<RequestReset />', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <RequestReset />
      </MockedProvider>
    )

    const form = wrapper.find('[data-test="request-reset"]')
    expect(toJSON(form)).toMatchSnapshot()
  })

  it('calls the mutation', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <RequestReset />
      </MockedProvider>
    )

    wrapper
      .find('input')
      .simulate('change', {
        target: { name: 'email', value: email }
      })

    wrapper
      .find('form')
      .simulate('submit')

    await wait()
    wrapper.update()

    expect(wrapper.text()).toContain('Success! Check your email')
  })
})