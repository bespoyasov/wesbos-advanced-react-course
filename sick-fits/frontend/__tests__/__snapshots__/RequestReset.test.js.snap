// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset /> renders and matches snapshot 1`] = `
Array [
  <Form
    data-test="request-reset"
    method="post"
    onSubmit={[Function]}
  >
    <form
      className="Form-sc-1xszr8q-0 krRxky"
      data-test="request-reset"
      method="post"
      onSubmit={[Function]}
    >
      <fieldset
        aria-busy={false}
        disabled={false}
      >
        <h2>
          Reset a password
        </h2>
        <DisplayError
          error={Object {}}
        />
        <label>
          Email
          <input
            name="email"
            onChange={[Function]}
            placehodler="Email"
            type="email"
            value=""
          />
        </label>
        <button
          type="submit"
        >
          Submit!
        </button>
      </fieldset>
    </form>
  </Form>,
  <form
    className="Form-sc-1xszr8q-0 krRxky"
    data-test="request-reset"
    method="post"
    onSubmit={[Function]}
  >
    <fieldset
      aria-busy={false}
      disabled={false}
    >
      <h2>
        Reset a password
      </h2>
      <DisplayError
        error={Object {}}
      />
      <label>
        Email
        <input
          name="email"
          onChange={[Function]}
          placehodler="Email"
          type="email"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Submit!
      </button>
    </fieldset>
  </form>,
]
`;
